lib/iolog/log_reader.go

```diff
@@ -0,0 +1,96 @@
+// Copyright 2017 VMware, Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package iolog
+
+import (
+	"encoding/binary"
+	"io"
+	"time"
+
+	log "github.com/Sirupsen/logrus"
+)
+
+// LogReader unwraps entires into LogEntry structs.
+type LogReader struct {
+	io.ReadCloser
+	prev []byte
+}
+
+// NewLogReader wraps an io.ReadCloser in a LogReader.
+func NewLogReader(r io.ReadCloser) *LogReader {
+	return &LogReader{ReadCloser: r}
+}
+
+// Read scans up to `\n` in the underlying stream, parses the header for that entry,
+// and copies the message into the supplied buffer
+func (lr *LogReader) Read(p []byte) (int, error) {
+	var (
+		err  error
+		n, w int
+	)
+	msg := lr.prev
+	if msg == nil {
+		// read a header
+		h := make([]byte, 10)
+		n = 0
+		for n < 10 {
+			w, err = lr.ReadCloser.Read(h[n:])
+			n += w
+			if err != nil {
+				break
+			}
+		}
+		if err != nil {
+			return 0, err
+		}
+		// parse header
+		ts := time.Unix(int64(binary.LittleEndian.Uint64(h[:8])), 0)
+		s := binary.LittleEndian.Uint16(h[8:10])
+		partial := (0x2&s)>>1 == 1
+		stream := int(0x1 & s)
+		size := int(s >> 2)
+		// read the associated entry
+		msg = make([]byte, size)
+		n = 0
+		for n < size {
+			w, err = lr.ReadCloser.Read(msg[n:])
+			n += w
+			if err != nil {
+				break
+			}
+		}
+		if err != nil && err != io.EOF {
```

Comment:
same here, you can check the condition in the loop and either return from there or break out and continue with the next statement